services_delegate.go

```diff
@@ -165,17 +180,36 @@ func (d *servicesDelegate) NotifyUpdate(node *memberlist.Node) {
 	log.Debugf("NotifyUpdate(): %s", node.Name)
 }
 
+// Try to pack as many messages into the packet as we can. Note that this
+// assumes that no messages will be longer than the normal UDP packet size.
+// This means that max message length is somewhere around 1398 when taking
+// messaging overhead into account.
 func packPacket(broadcasts [][]byte, limit int, overhead int) (packet [][]byte, leftover [][]byte) {
 	total := 0
-	leftover = make([][]byte, 0) // So we don't return unallocated buffer
-	for _, message := range broadcasts {
-		if total+len(message)+overhead < limit {
-			packet = append(packet, message)
-			total += len(message) + overhead
-		} else {
-			leftover = append(leftover, message)
+	lastItem := -1
+
+	// Find the index of the last item that fits into the packet we're building
+	for i, message := range broadcasts {
+		if total+len(message)+overhead > limit {
+			break
 		}
+
+		lastItem = i
+		total += len(message) + overhead
+	}
+
+	if lastItem < 0 && len(broadcasts) > 0 {
+		log.Error("All messages were too long to fit! No broadcasts sent!")
+		// There could be a scenario here where one hugely long broadcast could
+		// get stuck forever and prevent anything else from going out. There
+		// may be a better way to handle this. Scanning for the next message that
+		// does fit results in lots of memory copying and doesn't perform at scale.
+		return nil, broadcasts
 	}
 
-	return packet, leftover
+	// Save the leftover messages after the last one that fit. If this is too
+	// much, then set it to the lastItem.
+	firstLeftover := lastItem + 1
+
+	return broadcasts[:lastItem+1], broadcasts[firstLeftover:]
```

Comment:
Wouldn't `broadcasts[:firstLeftover]` be clearer than `broadcasts[:lastItem+1]`?